src/methods/to.js

```diff
@@ -669,7 +669,13 @@ export class ToImplementation {
         let resultType = result.getType();
 
         // ii. If Type(result) is not Object, return result.
-        if (Value.isTypeCompatibleWith(resultType, PrimitiveValue)) {
+        if (
+          Value.isTypeCompatibleWith(resultType, PrimitiveValue) ||
+          // If the result type is the top Value, then it might be an object,
```

Comment:
This seems like a significant semantic issue. Maybe only do this in pure mode?

Reply:
Not even in pure mode. The correct thing to do here is to report an error if result is an abstract value with the Top type.  Right now that is probably good enough. Eventually, BinaryExpression.computeBinary could catch the error and (in pure mode) return an abstract binary operation instead.

To put things another way: Returning an abstract value here that results in an expression that first calls method, then checks if the result is a primitive and alternatively either throws a runtime error or calls the other method, is pretty awkward to express and seems a silly way of doing what will happen naturally if we just emit the source level operator that caused this call to ToPrimitive.

Reply:
If I just report an error here, we don't solve the problem that we're hitting where we want to be able to leave these calls in place.

So I'll have to do the catching in computeBinary. That seems reasonable.

In general I want to make sure we don't unnecessarily leak too much by not considering the knowledge that we do have at this point. It's probably unlikely to matter for this case though.